python/cudf/cudf/utils/cudautils.py

```diff
@@ -235,6 +237,13 @@ def grouped_window_sizes_from_offset(arr, group_starts, offset):
     return window_sizes
 
 
+# This cache is keyed on the (signature, code, closure variables) of UDFs, so
+# it can hit for distinct functions that are similar. The lru_cache wrapping
+# compile_udf misses for these similar functions, but doesn't need to serialize
+# closure variables to check for a hit.
+_udf_code_cache = cachetools.LRUCache(maxsize=32)
+
+
 @lru_cache(maxsize=32)
```

Comment:
I could remove this, and just rely on the code cache - what would be a good way to evaluate the performance impact of doing so? My concern is that perhaps for a big UDF or one with lots of closure variables computing the key for the UDF code cache would be costly, but I don't know whether it makes any difference in an actual usage context.

Reply:
cc @brandon-b-miller maybe you have some ideas here since you've been exploring UDFs?

Reply:
cc @randerzander does your team have any large UDFs that we could use for performance testing here by any chance?

Reply:
"large UDFs" meaning using a significant number of state variables? I suppose just operating on a large number of input or output columns doesn't matter?

Reply:
There's two aspects to performance that I had in mind:

- The time taken to compute the key will scale with the size of the code, the size of any closure variables, and the number/type of arguments - so, UDFs that have a lot of code, operating on a lot of inputs / outputs, would be expected to spend more time evaluating whether there is a cache hit.
- The number of times that `compile_udf` gets called throughout execution - if it only gets used at startup and then the workflow proceeds using already-compiled code, then the performance impact of the cache won't be that relevant. If it's called in a tight loop (e.g. `for i in range(N): data.applymap(lambda x: x + 1)`) then the time spent computing cache keys might be more significant.

Reply:
It seems that removing the `@lru_cache` increases the time required to check the cache by around an order of magnitude (around the order of 1e-5ms from 1e-6ms) for some benchmarking with some device functions from FBPIC: https://gist.github.com/gmarkall/b3dc05580f24bdcfbfbc7fadd8c0b6cf (I picked these functions as some reasonably complex open source device functions I could find).

Given that leaving `@lru_cache` there doesn't really add much complexity, I'm inclined to leave this as-is, in case there are more complex use cases where the key computation would take a lot longer.

Reply:
@gmarkall my 2c: we're talking about ~5ns vs ~20ns here which doesn't make a different in the grand scheme in the pipeline. I'd say we should remove the `@lru_cache` to reduce complexity and just have a single caching mechanism instead.

Reply:
OK, I'll remove this shortly.

Reply:
@kkraus14 I've now removed this and updated the tests accordingly.

Reply:
Thanks @gmarkall looks good, resolving this.